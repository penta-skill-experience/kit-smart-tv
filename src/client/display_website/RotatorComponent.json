{
  "SWITCH_SPEED": 1000,
  "SWITCH_RATE": 15000
}
