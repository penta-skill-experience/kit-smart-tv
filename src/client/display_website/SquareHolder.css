#squareHeld {
    backdrop-filter: blur(25px);
}