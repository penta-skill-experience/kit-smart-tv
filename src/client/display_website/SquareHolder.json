{
  "SCROLL_REFRESH": 1000,
  "RATIO": 1.6,
  "ERROR_CLEAR_INTERVAL": 30000
}
