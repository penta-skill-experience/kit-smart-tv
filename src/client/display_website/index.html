<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Smart TV</title>
    <meta content="width=device-width, initial-scale=1" name="viewport">
    <link
            href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
            rel="stylesheet"
    />
</head>
<body>
<div id="root">
    <!-- react output will go inside here -->
</div>
</body>
</html>
