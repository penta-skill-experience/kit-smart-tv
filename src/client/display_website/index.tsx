import "./index.css";
import * as ReactDOM from "react-dom";
import * as React from "react";
import {RootComponent} from "./RootComponent";

ReactDOM.render(
    <RootComponent/>,
    document.getElementById("root")
);