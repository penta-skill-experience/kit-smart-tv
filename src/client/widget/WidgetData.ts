export class WidgetData {

    widgetId: string;
    location: number;
    fill: boolean;
    rawConfig: Object;

    constructor(widgetId: string, location: number, fill: boolean, rawConfig: Object) {
        this.widgetId = widgetId;
        this.location = location;
        this.fill = fill;
        this.rawConfig = rawConfig;
    }
}