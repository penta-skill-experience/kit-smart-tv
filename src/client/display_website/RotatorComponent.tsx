import * as React from "react";
import * as RotatorComponentConfig from "./RotatorComponent.json";
import {Carousel} from "react-responsive-carousel";
import "react-responsive-carousel/lib/styles/carousel.min.css";

export class RotatorComponent extends React.Component<any, any> {

    private pageSwitch() {
        const newIndex = (this.state.index + 1) % React.Children.count(this.props.children);
        this.setState({index: newIndex});
    }

    constructor(props) {
        super(props);
        this.state = {
            index: 0,
        };
        setInterval(() => this.pageSwitch(), RotatorComponentConfig.SWITCH_SPEED);
    }

    render() {
        return <div className="w-full h-full">
            <Carousel showIndicators={React.Children.count(this.props.children) > 1} showStatus={false} showArrows={false} autoPlay showThumbs={false} transitionTime={RotatorComponentConfig.SWITCH_SPEED} interval={RotatorComponentConfig.SWITCH_RATE} infiniteLoop={true} dynamicHeight={true}>
                {
                    React.Children.map(this.props.children, child =>
                        <div>
                            {child}
                        </div>)
                }
            </Carousel>
        </div>
    }
}