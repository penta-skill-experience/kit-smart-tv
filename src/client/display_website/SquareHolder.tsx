import * as React from "react";
import * as SquareHolderConfig from "./SquareHolder.json";
import "./SquareHolder.css";

interface SquareHolderProps {
    title: string;
    titleColor: string;
    accentColor: string;
    specialBoldFontColor: string;
    specialSubtleFontColor: string;
}

export class SquareHolder extends React.Component<SquareHolderProps, any> {

    pageScroll = function () {
        if (this.state.uniqueID === null || document.getElementById(this.state.uniqueID) === null) {
            return;
        }
        document.getElementById(this.state.uniqueID).scrollBy(0, this.state.scroll); // horizontal and vertical scroll increments
        if ((document.getElementById(this.state.uniqueID).scrollTop +
                document.getElementById(this.state.uniqueID).clientHeight) >=
            document.getElementById(this.state.uniqueID).scrollHeight || document.getElementById(this.state.uniqueID).scrollTop == 0) {
            this.state.scroll = -this.state.scroll;
        }
    }
    doesOverflow = function () {
        return document.getElementById(this.state.uniqueID).scrollHeight >= document.getElementById(this.state.uniqueID).clientHeight;
    }
    randomID = function () { //generate random ID to make scrolling behavior unique for each Squareholder
        let part = function () {
            return (Math.random().toString(10).substring(2));
        }
        return ("id" + part() + part() + part());
    };

    constructor(props) {
        super(props);
        this.state = {
            uniqueID: this.randomID(),
            hasError: false,
            error: undefined,
            scroll: SquareHolderConfig.SCROLL_SPEED
        };
        setInterval(() => this.pageScroll(), SquareHolderConfig.SCROLL_REFRESH);
        //(this.doesOverflow() ? setInterval(() => this.pageScroll(), 50):"");
    }

    static getDerivedStateFromError(error) {
        return {
            hasError: true,
            error: error,
        };
    }

    render() {
        return <div className = "box-border" style={{
            height: "45vh",
            width: "31vw",
            boxSizing: "border-box"
        }}>
            <div className={"w-full h-full shadow-2xl rounded-2xl"} id="squareHeld" style={{
                backgroundColor: this.props.accentColor,
                boxSizing: "border-box",
                overflow: "hidden"
            }}>
                <div
                    className={"sm:px-2 lg:px-3 xl:px-4 4xl:px-5 8xl:px-6 font-light leading-normal sm:text-base md:text-xl lg:text-2xl xl:text-4xl 2xl:text-5xl 8xl:text-6xl"}
                    style={{
                        color: this.props.titleColor
                    }}>
                    {this.props.title}
                </div>
                <div
                    className={"w-full h-full sm:pl-5 sm:pt-1 sm:pb-2 sm:pr-2 xl:pl-8 xl:pr-5 xl:pb-4 4xl:pl-12 overflow-x-scroll scrollbar-hide scroll-smooth overflow-hidden"}
                    id={this.state.uniqueID} style={{
                    scrollBehavior: "smooth"
                }}>
                    {this.state.hasError ?
                        // todo: make design for error message nicer
                        <p>Error while rendering widget: {this.state.error.message}</p>
                        : this.props.children  // only show children if they didn't produce an error
                    }
                </div>
            </div>
        </div>;
    }
}