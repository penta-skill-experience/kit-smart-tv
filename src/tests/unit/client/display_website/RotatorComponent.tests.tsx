import React from "react";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {configure, shallow} from "enzyme";
import toJson from "enzyme-to-json";
import {RotatorComponent} from "../../../../client/display_website/RotatorComponent";
import {WidgetData} from "../../../../client/widget/WidgetData";
import {SquareHolder} from "../../../../client/display_website/SquareHolder";
import {
    CafeteriaMenuDisplayComponent
} from "../../../../client/widget_catalog/cafeteria_menu/CafeteriaMenuDisplayComponent";
import {
    CafeteriaOpeningDisplayComponent
} from "../../../../client/widget_catalog/cafeteria_opening/CafeteriaOpeningDisplayComponent";

configure({adapter: new Adapter()});

const widgetOne = new WidgetData("cafeteria-menu", 2, {});
const widgetTwo = new WidgetData("cafeteria-opening", 2, {});

describe("rotatorComponent Snapshots", () => {

    test("rotatorComponent Snapshots without children", async () => {
        jest.useFakeTimers('legacy');
        Promise.resolve().then(() => jest.advanceTimersByTime(15000));
        jest.runOnlyPendingTimers();
        const wrapper = shallow(<RotatorComponent/>);
        await new Promise(process.nextTick);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test("rotatorComponent with children", async () => {
        jest.useFakeTimers('legacy');
        Promise.resolve().then(() => jest.advanceTimersByTime(15000));
        jest.runOnlyPendingTimers();
        const wrapper = shallow(<RotatorComponent>
            <SquareHolder displayComponentClass={CafeteriaMenuDisplayComponent}
                              rawConfig={widgetOne.rawConfig}
                              title={widgetOne.widgetId}
                              accentColor="black"
                              titleColor="black"
                              specialBoldFontColor="black"
                              specialSubtleFontColor="black">
             </SquareHolder>
             <SquareHolder displayComponentClass={CafeteriaOpeningDisplayComponent}
                          rawConfig={""}
                          title={widgetTwo.widgetId}
                          accentColor="black"
                          titleColor="black"
                          specialBoldFontColor="black"
                          specialSubtleFontColor="black">
            </SquareHolder>
            </RotatorComponent>
            );
        await new Promise(process.nextTick);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    afterEach(() => {
        jest.useRealTimers();
    });
    afterAll(() => {
        jest.restoreAllMocks();
    });
});