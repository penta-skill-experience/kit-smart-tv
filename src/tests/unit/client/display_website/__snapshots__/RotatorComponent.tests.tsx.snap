// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rotatorComponent Snapshots rotatorComponent Snapshots with design values and widget info 1`] = `
<div
  className="w-full h-full text-center"
>
  <div />
</div>
`;

exports[`rotatorComponent Snapshots rotatorComponent Snapshots without children 1`] = `
<div
  className="w-full h-full text-center"
>
  <div />
</div>
`;

exports[`rotatorComponent Snapshots rotatorComponent with children 1`] = `
<div
  className="w-full h-full text-center"
>
  <Carousel
    animationHandler="slide"
    autoPlay={true}
    axis="horizontal"
    centerSlidePercentage={80}
    dynamicHeight={false}
    infiniteLoop={true}
    interval={15000}
    labels={
      Object {
        "item": "slide item",
        "leftArrow": "previous slide / item",
        "rightArrow": "next slide / item",
      }
    }
    onChange={[Function]}
    onClickItem={[Function]}
    onClickThumb={[Function]}
    onSwipeEnd={[Function]}
    onSwipeMove={[Function]}
    onSwipeStart={[Function]}
    preventMovementUntilSwipeScrollTolerance={false}
    renderArrowNext={[Function]}
    renderArrowPrev={[Function]}
    renderIndicator={[Function]}
    renderItem={[Function]}
    renderThumbs={[Function]}
    selectedItem={0}
    showArrows={false}
    showIndicators={true}
    showStatus={false}
    showThumbs={false}
    statusFormatter={[Function]}
    stopOnHover={true}
    stopSwipingHandler={[Function]}
    swipeAnimationHandler={[Function]}
    swipeScrollTolerance={5}
    swipeable={true}
    transitionTime={1000}
    verticalSwipe="standard"
    width="100%"
  >
    <div
      key=".0"
    >
      <SquareHolder
        accentColor="black"
        displayComponentClass={[Function]}
        rawConfig={Object {}}
        specialBoldFontColor="black"
        specialSubtleFontColor="black"
        title="cafeteria-menu"
        titleColor="black"
      />
    </div>
    <div
      key=".1"
    >
      <SquareHolder
        accentColor="black"
        displayComponentClass={[Function]}
        rawConfig=""
        specialBoldFontColor="black"
        specialSubtleFontColor="black"
        title="cafeteria-menu"
        titleColor="black"
      />
    </div>
  </Carousel>
</div>
`;
